lib/geckodriver.js

```diff
@@ -0,0 +1,180 @@
+// transpile:main
+import events from 'events';
+import _ from 'lodash';
+import { logger } from 'appium-support';
+import { JWProxy } from 'appium-base-driver';
+import { SubProcess } from 'teen_process';
+import compareVersions from 'compare-versions';
+
+const log = logger.getLogger('Geckodriver');
+
+const LOG_DEBUG = 'debug';
+const DEFAULT_LOG_LEVEL = LOG_DEBUG;
+
+const DEFAULT_HOST = '127.0.0.1';
+const DEFAULT_PORT = 29146;
+const SYSTEM_GECKODRIVER = '/usr/local/bin/geckodriver';
+
+const MIN_SUPPORTED_VERSION = '0.26.0';
+const FIREFOX_PACKAGE = 'org.mozilla.firefox';
+const FIREFOX_MAIN_ACTIVITY = 'org.mozilla.gecko.BrowserApp';
+
+class Geckodriver extends events.EventEmitter {
+  constructor (args = {}) {
+    super();
+
+    const {
+      host = DEFAULT_HOST,
+      port = DEFAULT_PORT,
+      debug_level = DEFAULT_LOG_LEVEL
+    } = args;
+
+    this.proxyHost = host;
+    this.proxyPort = port;
+    this.jwproxy = new JWProxy({server: this.proxyHost, port: this.proxyPort, base: ''});
+    this.state = Geckodriver.STATE_STOPPED;
+    this.logLevel = debug_level;
+    this.verbose = this.logLevel === LOG_DEBUG;
+    this.capabilities = {};
+  }
+
+  parseVersion (stdout) {
+    return stdout
+      .match(/geckodriver\s+\d+\.\d+\.\d+/)[0]
+      .replace(/^geckodriver\s*/, '');
+  }
+
+  getVersion = _.memoize(async () => {
```

Comment:
this must be memoized per instance, not per class. See https://github.com/appium/appium-uiautomator2-driver/blob/97587ed63994fc7d445253d62456c2d897b4f15f/lib/driver.js#L144